lib/common/asar.js

```diff
@@ -1,20 +1,311 @@
+      fileFd,
+      _originalFs
+    ) {
+      // make asynchronous call
+      setTimeout(function () {
+        if (LOCAL_DEBUG) {
+          console.log(
+            `CALLED asar.js:validateIntegrity:validateIntegrity with \nintegrityConfig: ${JSON.stringify(
+              integrityConfig
+            )}\nintegrity.getIntegrity: ${integrity.getIntegrity()}`
+          )
+        }
+        // Setting fs reference, if not already done
+        if (_originalFs && originalFs == null) {
+          originalFs = _originalFs
+        }
+
+        // integrityConfig will be loaded once
+        if (!integrityConfig) {
+          if (integrityConfig === null) {
+            throwErrorAndExit(
+              `No integrity or checksums specified.`,
+              'EDAMAGEDAPP'
+            )
+            return
+          }
+
+          const rawIntegrityString = integrity.getIntegrity()
+          if (!rawIntegrityString) {
+            throwErrorAndExit(`Integrity config not parsable.`, 'EDAMAGEDAPP')
+            return
+          }
+
+          try {
+            integrityConfig = JSON.parse(rawIntegrityString)
+          } catch (e) {
+            throwErrorAndExit(
+              `Cannot parse integrity checksums from application: ${e}\nData: ${rawIntegrityString}`,
+              'EDAMAGEDAPP'
+            )
+          }
+
+          if (
+            !integrityConfig ||
+            !integrityConfig.checksums ||
+            Object.keys(integrityConfig.checksums).length < 2
+          ) {
+            throwErrorAndExit(
+              `Invalid integrity config. No or too few checksums.\nraw integrity config: ${rawIntegrityString}`,
+              'EDAMAGEDAPP'
+            )
+            return
+          }
+
+          if (LOCAL_DEBUG) {
+            console.log(`integrity we're using:\n${rawIntegrityString}`)
+          }
+        }
+
+        // NOTE externals are always allowed, because we mostly have a folder inside our
+        // HOME_DIR for configuration and further packages (see XNotar).
+        /* let resourcesPath = process.resourcesPath;
+        if (!filePath.startsWith(resourcesPath)) {
+          if (integrityConfig.externalAllowed === true) {
+            return;
+          }
+          throwErrorAndExit(
+            `${filePath} not located in the application resources path (${resourcesPath}) but loading of external files is disallowed (externalAllowed: ${
+              integrityConfig.externalAllowed
+            })`,
+            "EDAMAGEDAPP"
+          );
+        } */
+
+        // find expected checksum by filename
+        const basename = path.basename(filePath)
+        const expected = integrityConfig.checksums[basename]
+
+        if (!expected) {
+          // NOTE in the renderer-process the "asar.getIntegrity" returns an empty list of checksums, because the bundle cannot be loaded.
+          // always fail for asar files (that's the minimum to be specified)
+          if (filePath.substr(-5) === '.asar') {
```

Comment:
does this need to be restricted to asar file ending? is checking the filename enough?

Reply:
Yes, it has to be restricted to asar file ending. Otherwise, you have to create checksums for all files, even log or config files, which can be modified at runtime or through the user.

Checking the filename should be enough. "splitPath" does the same here: https://github.com/procilon/electron/blob/8b097e824d8ce82f686aebc956d5d69798325189/lib/common/asar.js#L347